numba/types/containers.py

```diff
@@ -511,6 +512,11 @@ def unify(self, typingctx, other):
         if isinstance(other, ListType):
             if not other.is_precise():
                 return self
+            mutable = self.mutable or other.mutable
```

Comment:
I think the unification should be:

```python
            mutable = self.mutable and other.mutable 
            if self.dtype == other.dtype:
                return ListType(self.dtype, mutable=mutable)
```

* result is mutable if and only if both input types are mutable
* the dtype must be the same

Reply:
Did that and all tests continue to pass so we will stick to that.